File: backend/app/static/levels.py
import json
import os
from flask import session
from .utils import green, no_spaces, app_folder, user_folder_path


def hint_handler(command, log):
    level = session['level']
    stage = session['stage']

    if level == 1:
        if stage == 1:
            return "Użyj komendy 'git merge friend_branch -m <wiadomość>'", ""
        if stage == 2:
            return "Pozbądź się konflitku a potem wpisz 'git add przepis.txt' i " \
                   "'git merge --continue' lub 'git commit -m <wiadomość>'. Pamiętaj, że musisz zachować dokładnie " \
                   "jedną z wersji przepisu (twoją, albo przyjaciela)", ""

    elif level == 2:
        if stage == 1:
            return "Użyj komendy 'git merge friend_branch -m <wiadomość>'", ""
        if stage == 2:
            return "Pozbądź się konflitku a potem wpisz 'git add style.json' i " \
                   "'git merge --continue' lub 'git commit -m <wiadomość>'. W pliku json " \
                   "style.json masz zostawić swoją wersję części 'header' i cudzą wersję 'footer'.", ""

    elif level == 3:
        if stage == 1:
            return "Użyj komendy 'git rebase liczby_catalana'", ""
        if stage == 2:
            return "Pozbądź się konflitku, a potem wpisz 'git rebasee --continue'\n" + \
                   "Pamiętaj, że na tym poziomie masz zostawić jedną z wersji funckji 'silnia' oraz " \
                   "dokończone implementacje funkcji 'taylor_e' oraz 'catalan'", ""

    elif level == 4:
        if stage == 1:
            return "Użyj komendy 'git cherry-pick <COMMIT>' gdzie commit to ciąg przynajmniej 4 znaków " \
                   "z hasza commita z komentarzem 'helpful defines'", ""

        if stage == 2:
            return "Pozbądź się konflitku, a potem wpisz 'git cherry-pick --continue'\n" + \
                   "Zaaplikuj wszystkie zmiany z drugiej gałęzi.", ""

    elif level == 5:
        if stage == 1:
            return "Użyj komendy 'git merge friend_branch -m <wiadomość>", ""
        if stage == 2:
            return "Użyj komendy 'git merge --abort", ""

    elif level == 6:
        if stage == 1:
            return "Użyj komendy 'git merge mat -m <wiadomość>", ""
        if stage == 2:
            return "Pozbądź się niepotrzebnych zmian z pliku wyjazd.txt, a następnie powtórz merge", ""

    elif level == 7:
        if stage == 1:
            return "Użyj komendy 'git merge rodzice -m <wiadomość>", ""
        if stage == 2:
            return "Konflikt jest dośc spory. I tak wiesz, że musisz się zgodzić na wszystkie zmiany rodziców "\
                   "więc najlepiej przerwać aktualnego merge za pomocą --abort i zrobić merge z flagą '-X theirs'", ""

    elif level == 8:
        if stage == 1:
            return "Możesz spróbować zrobić ten poziom na dwa sposoby. Albo po prostu używając od razu"\
                   "'git commit', albo najpierw wykonać checkout na gałąź dfsFix i tam 'git rebase master'", ""
        if stage == 2:
            return "Konflikt polega na tym, że na gałęzi master zamieniono nazwy zmiennych 'w' oraz 'x' na"\
                   "odpowiednio 'startowy' i 'aktualny'. Zaaplikuj zmianę z dfsFix z nowymi zmiennymi!", ""

    return "", "ERROR"


def check_stage(log):
    level = session['level']
    stage = session['stage']

    if 1 <= level <= 5 or level == 7 or level == 8:
        if stage == 1 and 'conflict' in log:
            session['stage'] = 2
            session.modified = True

    elif level == 6:
        if log['stderr'].startswith('error: Your local changes to the'):
            session['stage'] = 2
            session.modified = True

    else:
        pass


def add_extra_allowed(extra_allowed):
    # Do ustalenia jest jeszcze to na jakie komendy pozwalamy na danych levelach

    level = session['level']
    stage = session['stage']

    if level == 1 or level == 2:
        extra_allowed.append('git add')
        extra_allowed.append('git merge')
        if stage == 2:
            extra_allowed.append('git commit')

    elif level == 3:
        extra_allowed.append('git rebase')
        extra_allowed.append('git add')

    elif level == 4:
        extra_allowed.append('git cherry-pick')
        extra_allowed.append('git add')

    elif level == 5:
        extra_allowed.append('git merge')

    elif level == 6:
        extra_allowed.append('git merge')
        extra_allowed.append('git add')

    elif level == 7:
        extra_allowed.append('git merge')
        extra_allowed.append('git add')

    elif level == 8:
        extra_allowed.append('git merge')
        extra_allowed.append('git rebase')
        extra_allowed.append('git add')
        extra_allowed.append('git branch')
        extra_allowed.append('git checkout')

    else:
        pass  # TODO level 6, 7 i 8.


def check_success(log):
    # W tym momencie zakładamy, że dokonał się merge i drzewa są takie same
    # czasami chcemy jednak sprawdzić, czy wartość niektórych plików jest
    # taka, jak tego oczekujemy od użytkowników

    level = session['level']
    level_directory = os.path.join(app_folder(), 'levels', f'level{level}')
    if level == 1:
        with open(os.path.join(level_directory, 'friend_file'), 'r') as f:
            expected1 = no_spaces(f.read())
        with open(os.path.join(level_directory, 'your_file'), 'r') as f:
            expected2 = no_spaces(f.read())

        try:
            with open(os.path.join(user_folder_path(), 'przepis.txt'), 'r') as f:
                user_output = no_spaces(f.read())
        except FileNotFoundError:
            log['reset'] = 'Nie ma pliku przepis.txt'
            return

        print(f"{expected1 = }")
        print(f"{expected2 = }")
        print(f"{user_output =}")

        if expected1 == user_output or expected2 == user_output:
            log['success'] = True
        else:
            log['reset'] = "Zawartość pliku 'przepis.txt' niezgodna z poleceniem"

    elif level == 2:
        with open(os.path.join(level_directory, 'expected_answer.json')) as f:
            expected_output = json.load(f)

        try:
            with open(os.path.join(user_folder_path(), 'style.json')) as f:
                try:
                    user_output = json.load(f)
                except json.JSONDecodeError:
                    log['reset'] = "Zawartość pliku 'style.json' nie jest w poprawnym formacie JSON!"
                    return

        except FileExistsError:
            log['reset'] = "Nie ma pliku 'style.json'"
            return False

        if expected_output != user_output:
            log['reset'] = "Oczekiwano innej zawartości pliku 'style.json'"
            return

        log['success'] = True

    elif level == 3:
        ok = False
        try:
            with open(os.path.join(user_folder_path(), 'kod.py'), 'r') as f:
                user_output = no_spaces(f.read())
        except FileNotFoundError:
            log['reset'] = 'Nie ma pliku kod.py'
            return

        for i in range(4):
            with open(os.path.join(level_directory, f'expected_output{i}.py'), 'r') as f:
                expected_output = no_spaces(f.read())
                if expected_output == user_output:
                    ok = True
        if ok:
            log['success'] = True
        else:
            log['reset'] = "Niepoprawna zawartośc pliku kod.py"

    elif level == 4:
        try:
            with open(os.path.join(user_folder_path(), 'kod.cpp'), 'r') as f:
                user_output = no_spaces(f.read())
        except FileNotFoundError:
            log['reset'] = 'Nie ma pliku kod.cpp'
            return

        with open(os.path.join(level_directory, f'friend_kod2.cpp'), 'r') as f:
            expected_output = no_spaces(f.read())

        if user_output == expected_output:
            log['success'] = True
        else:
            log['reset'] = "Niepoprawna zawartośc pliku kod.cpp"

    elif level == 5:
        pass  # sprawdzenia dokonujemy wcześniej w module commands.py

    elif level == 6:
        out_file = 'wyjazd.txt'
        try:
            with open(os.path.join(user_folder_path(), out_file), 'r') as f:
                user_output = no_spaces(f.read())
        except FileNotFoundError:
            log['reset'] = f'Nie ma pliku {out_file}'
            return

        with open(os.path.join(level_directory, f'comb.txt'), 'r') as f:
            expected_output = no_spaces(f.read())

        if user_output == expected_output:
            log['success'] = True
        else:
            log['reset'] = f"Niepoprawna zawartośc pliku {out_file}"

    elif level == 7:
        out_files = ['ogrod.txt',
                     os.path.join('parter', 'kuchnia.txt'),
                     os.path.join('parter', 'salon.txt'),
                     os.path.join('pietro', 'lazienka.txt'),
                     os.path.join('pietro', 'garderoba.txt'),
                     os.path.join('pietro', 'sypialnia.txt'),
                     ]

        for file in out_files:
            try:
                with open(os.path.join(user_folder_path(), file), 'r') as f:
                    user_output = no_spaces(f.read())
            except FileNotFoundError:
                log['reset'] = f'Nie ma pliku {file}'
                return

            with open(os.path.join(level_directory, 'ogrod', file), 'r') as f:
                expected_output = no_spaces(f.read())

            if user_output != expected_output:
                log['reset'] = f"Niepoprawna zawartośc pliku {file}"
                return

            log['success'] = True

    elif level == 8:
        try:
            with open(os.path.join(user_folder_path(), 'kod.py'), 'r') as f:
                user_output = no_spaces(f.read())
        except FileNotFoundError:
            log['reset'] = f'Nie ma pliku kod.py'
            return

        with open(os.path.join(level_directory, 'expected_output.py'), 'r') as f:
            expected_output = no_spaces(f.read())

        if user_output != expected_output:
            print(f"{user_output}")
            print(f"{expected_output}")
            log['reset'] = f"Niepoprawna zawartośc pliku 'kod.py'"
            return

        log['success'] = True

    else:
        log['reset'] = "Przeszedłeś wszystkie poziomy!"
        return

File: frontend/solve-conflicts/src/components/levels/Levels.ts
export const levels = [
  [
    {
      message: `Witaj w solveConflicts! Mamy nadzieję, że
      wspólnie nauczymy się czegoś ciekawego o rozwiązywaniu
      konfliktów w systemie kontroli wersji GIT.`,
      width: '600px',
      height: '200px',
      stage: 1,
    },
    {
      message: `Razem ze swoim kolegą planujecie umieścić na swojej stronie
      internetowej przepis na naleśniki. W tym celu każdy z Was
      napisał własną wersję tego jakich proporcji należy użyć do ciasta.
      Musisz teraz zadecydować, który przepis jest lepszy i zostanie opublikowany.
      Rozpocznij od wpisania komendy: ‘git branch‘.`,
      width: '600px',
      height: '300px',
      stage: 1,
    },
    {
      message: `Zauważ, że znajdujesz się aktualnie na gałęzi ‘master’, gdzie
      jest plik z Twoją wersją przepisu. Twój kolega ma przepis na osobnej
      gałęzi (friend branch). Spróbuj automatycznie połączyć te dwie
      gałęzie przy użyciu komendy: git merge friend_branch”
      W razie potrzeby pamiętaj o komendach: ‘help’ oraz ‘hint’.`,
      width: '600px',
      height: '300px',
      stage: 1,
    },
    {
      message: `Niestety… zawartości plików są istotnie różne i GIT nie jest
      w stanie ich automatycznie połączyć. Zauważ, że zawartość pliku
      ‘przepis.txt’ uległa zmianie. Znajdują się w nim miejsce z dwoma wyraźnie
      oddzielonymi sekcjami. Sekcja HEAD oznacza wersję Twojej gałęzi. Druga
      sekcja to zmiany kolegi. Twoim zadaniem jest zmodyfikować ten plik wybierając
      jedną z dwóch wersji, po czym użyć komendy ‘git add’ a następnie dokończyć merge.`,
      width: '600px',
      height: '300px',
      stage: 2,
    },
  ],
  [
    {
      message: `Pierwszy poziom był dość prosty. Teraz będzie nieco trudniej.
      Masz za zadanie połączyć dwie wersje pliku ‘style.json’ z konfiguracją
      stylowania Waszej strony internetowej. W tym pliku są trzy sekcje (“header”,
      “main-table” oraz “footer”). Rozpocznij przejrzenia zawartości pliku oraz
      próby zmergowania tych gałęzi!`,
      width: '600px',
      height: '250px',
      stage: 1,
    },
    {
      message: `O ile obie wersje pliku mają taką samą wersję
      sekcji “main-table”, to pozostałe sekcje się od siebie różnią. Wspólnie
      z kolegą ustaliliście, aby zachować Twoją wersję kawałka odpowiadającego
      za sekcję “footer”, ale wersję kolegi jeżeli chodzi o “header”.
      Do dzieła!`,
      width: '600px',
      height: '250px',
      stage: 1,
    },
  ],
  [
    {
      message: `Całkiem nieźle! Udało Ci się rozwiązać już dwa konflikty. Ale czy
      konflikt może wystąpić jedynie przy próbie wykonania ‘git merge’? Okazuje
      się, że konflikty mogą pojawiać się również przy innych komendach
      zmieniających strukturę repozytorium.`,
      width: '600px',
      height: '250px',
      stage: 1,
    },
    {
      message: `Jedną z nich jest git rebase. Pomoże nam ono uporządkować historię
      naszego repozytorium. Pracowałeś ostatnio nad dwiema gałęźmi, z których jedna
      wyliczała liczby Catalana, a na drugiej umieściłeś kod szeregu Taylora.
      Postanowiłeś, że możesz połączyć
      ten plik w jeden i sprawić żeby historia wyglądała tak, jakbyś nigdy nie rozdzielał pracy na dwoje.
      Spróbuj wykonać komendę ‘git rebase liczby_catalana’.`,
      width: '600px',
      height: '300px',
      stage: 1,
    },
    {
      message: `Funkcja silnia posiada dwie różne implementacje. Wybierz tę, z której wolałbyś korzystać, a następnie ‘pogódź’ resztę pliku zachowując ukończone
      implementacje pozostałych dwóch funkcji. Po wszystkim użyj komendy ‘git rebase’ z flagą ‘–continue’.`,
      width: '600px',
      height: '150px',
      stage: 2,
    },
  ],
  [
    {
      message: `Razem z kilkoma znajomymi postanowiliście wystąpić w drużynowym konkursie
      programistycznym. Polega on na tym, że każdy próbuje wymyślić swoją
      strategię na rozwiązanie problemu optymalizacyjnego, a po jakimś czasie
      i naradach drużyna przechodzi w tryb implementacji kodu.`,
      width: `600px`,
      height: `200px`,
      stage: 1,
    },
    {
      message: `Wpadłeś na bardzo skomplikowany, ale obiecujący pomysł. Niestety
      wymyślenie go zabrało Ci trochę czasu. Chcesz go nadrobić. Pomyślałeś
      więc, że skorzystasz z szablonu kodu, który w międzyczasie zdążył już
      napisać jeden z Twoich kolegów. Wykonał on już kilka komitów na gałęzi
      swojego kodu, ale dla Ciebie
      istotny jest tylko jeden z nich o nazwie “defines”. W tym komicie jest
      kilka przydatnych linii kodu, których nie ma sensu przepisywać
      automatycznie.`,
      width: '600px',
      height: '300px',
      stage: 1,
    },
    {
      message: `Za pomocą “git cherry-pick COMMIT” wyłuskaj komit, którego
      potrzebujesz i dołącz go do swojej gałęzi, a jeżeli w międzyczasie pojawi
      się również jakiś konflikt, to go rozwiąż!`,
      width: '600px',
      height: '150px',
      stage: 1,
    },
    {
      message: `A niech to! Znowu jeden z tych przeklętych konfliktów! Na pewno dasz sobie z nim radę.`,
      width: '600px',
      height: '120px',
      stage: 2,
    },
    {
      message: `Może wydawać się to dziwne, ale jedna z sekcji powstałych w pliku
      konfliktowym jest pusta, ale oznacza, że git oznacza brak zmian, jako zmianę.`,
      width: '600px',
      height: '150px',
      stage: 2,
    },
  ],
  [
    {
      message: `Ten poziom, będzie krótką odskocznią od rozwiązywania konfliktów, ale
      też niesie w sobie nieco przydatnej wiedzy. Wyobraź sobie sytuację, gdzie
      w pracy dostałeś zlecenie na wykonanie połączenie dwóch (lub więcej)
      bardzo rozbudowanych gałęzi. Rozwiązałeś już część konfliktów, ale
      w trakcie pracy, Twój szef powiedział, żeby się jednak wstrzymać
      z mergowaniem, bo nie jest jeszcze do końca pewien, które wersje plików
      chciałby zostawić.`,
      width: '600px',
      height: '300px',
      stage: 1,
    },
    {
      message: `Jedzie na urlop i wróci dopiero za tydzień, więc nie ma co
      bezczynnie czekać. W tej sytuacji należy zrezygnować z aktualnego stanu
      i próbować wrócić do prac na jednej z gałęzi. Jak to zrobić? Okazuje się,
      że sprawa jest bardzo prosta. Wystarczy wykonać komendę “git merge --abort”,
      aby repozytorium powróciło do stanu sprzed wywołania komendy merge.`,
      width: '600px',
      height: '300px',
      stage: 1,
    },
    {
      message: `Okazuje się, że flaga --abort działa
      w bardzo podobny sposób, jeżeli chcielibyśmy cofnąć zmiany w trakcie
      ‘git rebase’ albo ‘git cherry-pick’`,
      width: '600px',
      height: '200px',
      stage: 1,
    },
  ],
  [
    {
      message: `Razem z bratem planujecie wyjechać latem w góry pozdobywać Tatrzańskie
      szczyty. W tym celu postanowiliście zrobić kalendarz wypraw na szlaki.
      Każdy z Was podał jakieś propozycje tras, i teraz pora na połączenie
      pomysłów. Użyj komendy git merge!`,
      width: '600px',
      height: '200px',
      stage: 1,
    },
    {
      message: `Niestety okazuje się, że merge się nawet nie rozpoczął… bo z pliku
      ‘wyjazd.txt’ znalazła się jakaś przypadkowa zmiana od ostatniego komita.
      W takiej sytuacji najlepiej byłoby pozbyć się tych niepotrzebnych zmian
      i wrócić do sytuacji bezpośrednio po zrobieniu ostatniego komita. Na
      szczęście Git daje nam taką możliwość! Z pomocą przychodzi komenda git restore.`,
      width: '600px',
      height: '300px',
      stage: 2,
    },
    {
      message: `Użyj tej komendy by, przywrócić prawidłowy stan dla pliku wyjazd.txt,
      a następnie połącza gałęzie ‘pat’ i ‘mat’ (jeżeli zajdzie konflikt, to
      zadbaj o jego rozwiązanie)`,
      width: '600px',
      height: '200px',
      stage: 2,
    },
  ],
  [
    {
      message: `Postanowiłeś zrobić swoim rodzicom prezent! Na ich trzydziestą rocznicę
      ślubu chcesz zrobić im remont mieszkania. Nic prostszego… Już nawet
      przygotowałeś arkusz, w którym wpisałeś co chcesz im wyremontować.`,
      width: '600px',
      height: '200px',
      stage: 1,
    },
    {
      message: `Jak to jednak w życiu bywa… rodzice po tym jak usłyszeli, że chcesz
      zrobić im prezent, postanowili wprowadzić pewne poprawki do arkusza…
      Stworzyli nawet nową gałąź w systemie GIT by tam notować swoje zmiany.`,
      width: '600px',
      height: '300px',
      stage: 1,
    },
    {
      message: `Poprosili Cię o uwzględnienie tych zmian… Sytuacja jest opłakana, ilość
      poprawek, które wprowadzili rodzice jest zatrważająca. Trudno Ci będzie w
      takiej sytuacji odmówić. Nie wiemy jak uratować Twój portfel, ale możemy
      pokazać Ci metodę na zaoszczędzenie czasu.`,
      width: '600px',
      height: '250px',
      stage: 1,
    },
    {
      message: `Dla komendy merge można podać flagę ‘-X’, która pozwala na określenie
      strategii przy merdżowaniu. i tak gdy podamy ‘-X ours’ to wszystkie
      konflikty będą rozstrzygane na korzyść HEAD, a dla ‘-X theirs’ na korzyść drugiej gałęzi.`,
      width: '600px',
      height: '200px',
      stage: 1,
    },
    {
      message: `W tym poziomie, możesz albo spróbować wykonać wszystkie zmiany ręcznie
      (zawsze zostawiając wersje z gałęzi rodziców), albo po prostu skorzystać z flagi -X`,
      width: '600px',
      height: '150px',
      stage: 1,
    },
  ],
  [
    {
      message: `Prowadzisz prace nad projektem programistycznym na gałęzi ‘develop’.
      Na potrzeby naprawy kawałka kodu stworzyłeś gałąź bugFix i dokonałeś
      na niej kilka zmian. Twoim zadaniem jest podłączenie tych zmian
      do głównego nurtu gałęzi. Możesz to zrobić za pomocą ‘merge’, albo ‘rebase’.`,
      width: '600px',
      height: '200px',
      stage: 1,
    },
    {
      message: `
      A jednak postał konflikt. Rozwiąż go i odbierz swój wymarzony dyplom ukończenia kursu!.`,
      width: '600px',
      height: '120px',
      stage: 2,
    },
  ],
];

export const tasks = [
  `W repozytorium znalazły się dwie różne wersje przepisu na
  naleśniki. Twoim zadaniem jest połączenie dwóch gałęzi wybierając jedną z wersji. Do rozwiązania zadania
  należy użyć komendy ‘git merge’.`,
  `Twoim zadaniem jest połączenie gałęzi z dwoma wersjami tego samego pliku
  ze stylami. Dokonaj tego w taki sposób, by po połączeniu w pliku została
  Twoja wersja sekcji “footer” oraz cudza wersja “header”.
  Do wykonania zadania użyj komendy ‘git merge’.`,
  `Nie tylko ‘merge’ powoduje konflikty. Spróbuj użyć ‘git rebase’ do
  przeniesienia swoich zmian pliku kod.py do gałęzi ‘liczby_catalana’. Napotkasz
  konflikt, który trzeba rozwiązać wybierając jedną z dostępnych
  implementacji funkcji ‘silnia’ oraz dokończone implementacje pozostałych
  funkcji.`,
  `Na tym poziomie chcemy zademonstrować działanie komendy “git cheery-pick”
  oraz to, że może ona powodować konflikty. Za jej pomocą dostań się do
  kodu zawartego w komicie “przydatne dyrektywy”. Skorzystaj z całego kodu
  oferowanego przez ten commit.`,
  `Czasami porzucenie zamiaru połączenia gałęzi może być najlepszym wyborem.
  Zwłaszcza gdy nie jest się pewny tego jak rozwiązać konflikty,
  lub jest ich zbyt wiele do rozwiązania w danym momencie. W takim przypadku
  pomocna może się okazać flaga “--abort” komendy “git merge”. Rozpocznij
  proces ‘merge’, a następnie go przerwij.`,
  `Bywa tak, że merge nie rozpocznie się, ponieważ lokalnie są obecne pliki,
  które nie zostały dodane do ‘przechowalni’ (staging area). Tak stało się
  również z plikiem ‘wyjazd.txt’. Sprawdź czy po użyciu ‘git restore’, które
  odrzuci nieskomitowane zminy z tego pliku, sytuacja ulegnie zmianie.`,
  `Jeżeli wiemy, że konflikty będą rozwiazywane zawsze na korzyść jednej z
  gałęzi, to możemy podać do ‘git merge’ flagę ‘-X’, która określi
  strategię automatycznego rozwiązywania konfliktów. W tym przypadku
  podanie argumentu ‘theirs’ spowoduje, że przyjmiemy domyślnie
  zmiany zawarte w gałęzi, które dołączamy do HEAD.`,
  `Czasami istnieje więcej niż jedna droga komend gitowych, która prowadzi
  nas to oczekiwanego efektu. W zależności od gustu, część programistów
  preferuje używanie ‘merge’, a część ‘rebase’ (co zmienia kształt
  grafu historii repozytorium). Twoim zadaniem jest zaaplikować zmiany
  z gałęzi odpluskwiającej kod, zachowując jednocześnie zmiany nazewnictwa
  dwóch zmiennych ‘w’ i ‘x’ na bardziej czytelne odpowiedniki.`,
];
